Use 'fileprivate' for synthesized members of 'private' decls

jrose-apple opened the pull request:
Since `private` means "limit to the enclosing scope (and extensions thereof)", putting it on a member means that the member can't be accessed everywhere the type might show up. That's normally a good thing, but it's not the desired effect for synthesized members used for derived conformances, and when it comes to class initializers this actually violates AST invariants.

rdar://problem/39478298

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

itaiferber commented:
How does this affect the synthesis of `CodingKeys`? We have always explicitly synthesized `private enum CodingKeys` on purpose; exposing them to the rest of the file isn't necessarily something we want. (Or at least, this is something we need to consider when changing.)

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2259,9 +2259,21 @@ ValueDecl::getFormalAccessScope(const DeclContext *useDC,
 
-void ValueDecl::copyFormalAccessFrom(ValueDecl *source) {
+void ValueDecl::copyFormalAccessFrom(const ValueDecl *source,
+                                     bool sourceIsParentContext) {
```
Do you need sourceIsParentContext, or can you just check if the source's parent DeclContext is a SourceFile, and if so, turn Private to FilePrivate?

jrose-apple replied:
It's relevant for any type, unfortunately, not just those at the top level. If we wanted to do it without the flag we could check if `source` and `this` are related in some way, but that seems more complicated for something every call site already knows.

jrose-apple committed:
Use 'fileprivate' for synthesized members of 'private' decls

Since 'private' means "limit to the enclosing scope (and extensions
thereof)", putting it on a member means that the member can't be
accessed everywhere the type might show up. That's normally a good
thing, but it's not the desired effect for synthesized members used
for derived conformances, and when it comes to class initializers this
actually violates AST invariants.

rdar://problem/39478298

slavapestov reviewed (approved):
The fix makes sense, you can merge it as-is but I have an optional suggestion that may not make sense

jrose-apple commented:
Made the tweaks mentioned above.

@swift-ci Please smoke test
